docs/hub/how-to-upstream.md

```diff
@@ -0,0 +1,263 @@
+
+```python
+>>> repo = Repository(local_dir="huggingface-hub", clone_from="<user>/<dataset_id>", revision='branch1')
+>>> repo.git_checkout("branch2")
+```
+
+### Add
+
+Add a file to commit with `git_add`:
+
+```python
+>>> repo = Repository(local_dir="huggingface-hub", clone_from="<user>/<dataset_id>")
+>>> repo.git_add("my-awesome-file")
+```
+
+This method has an `auto-lfs-track` parameter that can be set to `True` if you want to automatically track large files (>10MB):
+
+```python
+>>> repo = Repository(local_dir="huggingface-hub", clone_from="<user>/<dataset_id>")
+>>> repo.git_add("my-awesome-file", auto_lfs_track=True)
+```
+
+### Commit
+
+Commit a file with `git_commit`:
+
+```python
+>>> repo = Repository(local_dir="huggingface-hub", clone_from="<user>/<dataset_id>")
+>>> repo.git_add(commit_message="Commit my-awesome-file to the Hub")
+```
+
+### Pull
+
+Update a current local branch with `git_pull`:
+
+```python
+>>> repo = Repository(local_dir="huggingface-hub", clone_from="<user>/<dataset_id>")
+>>> repo.git_pull()
+```
+
+Set `rebase=True` if you want your local commits to occur after your branch is updated with the new commits from the remote:
+
+```python
+>>> repo = Repository(local_dir="huggingface-hub", clone_from="<user>/<dataset_id>")
+>>> repo.git_pull(rebase=True)
+```
+
+### Push
+
+Push all your local branch commits to the remote branch with `git_push`:
+
+```python
+>>> repo = Repository(local_dir="huggingface-hub", clone_from="<user>/<dataset_id>")
+>>> repo.git_push()
+```
+
+Change the `blocking` parameter to `False` if you would like to push your commits asynchronously. Non-blocking behavior is helpful when you want to continue running your script while you push your commits.
+
+```python
+>>> repo = Repository(local_dir="huggingface-hub", clone_from="<user>/<dataset_id>")
+>>> repo.git_push(blocking=False)
+```
+
+Check the status of your push with the `command_queue` property:
+
+```python
+>>> last_command = repo.command_queue[-1]
+>>> last_command.status
+# -> -1 indicates the push is ongoing.
+# -> 0 indicates the push has completed successfully.
+# -> Non-zero code indicates the error code if there was an error.
+```
+
+When `blocking=False`, commands are tracked and your script will only exit when all pushes are completed, even if other errors occur in your script. Some additional useful commands for checking the status of a push include:
+
+```python
+# Inspect an error.
+>>> last_command.stderr
+
+# Check whether a push is completed or ongoing.
+>>> last_command.is_done
+
+# Check whether a push command has errored.
+>>> last_command.failed
+```
+
+### `push_to_hub`
+
+For convenience, the `Repository` class also has a `push_to_hub` method that will add files to commit, commit the files, and push them to a repository:
+
+```python
+>>> repo = Repository(local_dir="huggingface-hub", clone_from="<user>/<dataset_id>")
+>>> repo.push_to_hub(commit_message="Commit my-awesome-file to the Hub")
+```
```

Comment:
I think this entire part is great and goes in detail about possible methods - but it omits the `commit` context manager, which is arguably the simplest and less error-prone method. If one method was to be presented, I'd vote for the `commit` context manager above all as it does everything (pull, add, commit, push) and it's hard to mess up (contrary to the `push_to_hub` method, which is less intuitive imo, as it requires a pull first, then saving the files, then calling the method).